src/app/convai-checker.component.spec.ts

```diff
@@ -1455,4 +1500,394 @@ describe('Convai checker test', () => {
+           expectedShape = Shape.SQUARE;
+         } else {
+           expectedShape = Shape.CIRCLE;
+         }
+         verifyLoadingWidgetHasShape(checker, expectedShape);
+
+         if (callCount < 2) {
+           callCount++;
+           // Fire another request.
+           setTextAndFireInputEvent(queryTexts[callCount], textArea);
+         }
+       });
+    });
+
+    verifyCircleSquareDiamondWidgetVisible();
+
+    let textArea = fixture.debugElement.query(
+      By.css('#' + checker.inputId)).nativeElement;
+
+    // Send an input event to trigger the service call.
+    setTextAndFireInputEvent(queryTexts[callCount], textArea);
+  }));
+
+  it('Test emoji change for score thresholds', async(() => {
+    let fixture = TestBed.createComponent(ConvaiCheckerCustomDemoSettingsTestComponent);
+
+    // Configure settings.
+    let demoSettings = getCopyOfDefaultDemoSettings();
+    demoSettings.scoreThresholds = [0, 0.6, 0.8];
+    demoSettings.loadingIconStyle = LoadingIconStyle.EMOJI;
+    fixture.componentInstance.setDemoSettings(demoSettings);
+
+    fixture.detectChanges();
+    let checker = fixture.componentInstance.checker;
+
+    // Set up the mock responses for the series of three requests that will be
+    // made in the test.
+    let queryTexts = [
+      'Your mother was a hamster',
+      'Your father smelled of elderberries',
+      'What is the air velocity of an unladen swallow?'
+    ];
+
+    let callCount = 0;
+    let mockResponses = [
+      getMockCheckerResponseWithScore(0.9, checker.getToken(queryTexts[0])),
+      getMockCheckerResponseWithScore(0.7, checker.getToken(queryTexts[1])),
+      getMockCheckerResponseWithScore(0.2, checker.getToken(queryTexts[2]))
+    ];
+
+    let mockBackend = TestBed.get(MockBackend);
+    mockBackend.connections
+     .subscribe((connection: MockConnection) => {
+       fixture.detectChanges();
+       // Check the UI state before returning the repsonse.
+
+       verifyEmojiWidgetVisible();
+
+       // TODO(rachelrosen): Figure out how to "fast-forward" the state of the
+       // animation in the test environment without messing up the other test
+       // state, and then uncomment the code below to check the opacity of the
+       // emoji icons here.
+       // verifyEmojiIconsInDomWithZeroOpacity();
+
+       expect(checker.statusWidget.isLoading).toBe(true);
+
+       connection.mockRespond(
+         new Response(
+           new ResponseOptions({
+              body: mockResponses[callCount]
+           })
+         )
+       );
+
+       // Wait for async code to complete.
+       fixture.whenStable().then(() => {
+         fixture.detectChanges();
+         // Checks the UI state after the response has been received.
+
+         // Checks that loading has stopped.
+         expect(checker.statusWidget.isLoading).toBe(false);
+         expect(checker.statusWidget.isPlayingLoadingAnimation).toBe(false);
+
+         verifyEmojiWidgetVisible();
```

Comment:
I noticed that this gets called a lot more often than the `verifyCircleSquareDiamondWidgetVisible` method gets called...is there a reason for that?

Reply:
There was a discrepancy of 1, and that's because I missed a call to verifyCircleSquareDiamondWidgetVisible. Thanks.